`@0x/contracts-broker`: Property-based Gods Unchained orders

mzhu25 opened the pull request:
## Description
Implements the Broker and GodsUnchainedValidator contracts, as described in https://github.com/0xProject/ZEIPs/issues/75
Also moves `LibAssetDataTransfer` from `@0x/contracts-exchange-forwarder` to `@0x/contracts-exchange-libs`. 

A couple of things to note:
- The Broker currently doesn't support orders with taker fees. This is because, like the Forwarder, the Broker serves as the taker of the order, so the taker fee asset would need to be transferred to the Broker prior to the fill (more approvals to be set by the actual taker, etc.). You can't mix and match property-based and vanilla orders in `batchBrokerTrade` for the same reason.
- Could add `msg.sender` checks to the fallback and `safeBatchTransferFrom`. 
- The way that the `takerAssetAmount` and the nested 1155 amounts work together is somewhat interesting for brokered assets. Refer to the comment above `encodeBrokerAssetData` for details
- Right now `safeBatchTransferFrom` is fairly opinionated in that it enforces that `from == address(this)` and assumes the brokered asset is 721. Generalizing this could open up a few interesting use cases, but this can wait for a later iteration. 

<!--- Describe your changes in detail -->

## Testing instructions
Unit tests for the GodsUnchainedValidator contract are found in the new `@0x/contracts-broker` package, while Broker <> GodsUnchainedValidator integration tests are in `@0x/contracts-integrations`

<!--- Please describe how reviewers can test your changes -->

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

<!-- * New feature (non-breaking change which adds functionality) -->

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

mzhu25 commented on contracts/broker/contracts/src/Broker.sol:
```diff
@@ -0,0 +1,253 @@
+
+            // Revert with returned data if staticcall is unsuccessful
+            if (!success) {
+                assembly {
```
we need this to be `staticcall` though, right?

dorothy-zbornak replied:
Solidity should compile to that because `checkBrokerAsset()` is declared `view`.

mzhu25 commented on contracts/broker/contracts/src/Broker.sol:
```diff
@@ -0,0 +1,253 @@
+    )
+        external
+    {
+        if (from != address(this)) {
```
I put this in to handle the case where e.g. both `makerFeeAssetData = takerAssetData`. Without this check, the Broker would send one of the taker's assets to the fee recipient, but this would be disguised as a maker fee

dorothy-zbornak replied:
Could this also be an issue for the taker fee? There's nothing explicitly preventing filling an order w/ a taker fee right?

mzhu25 replied:
Yeah, I suppose so, but at least if the `takerFee` is non-zero the taker should expect to be sending _something_ to the maker, so it's kinda on them/the front-end to surface what that thing is. Also I can't really think of an easy way for the Broker to distinguish between a taker -> maker transfer vs a taker -> feeRecipient transfer

dorothy-zbornak commented on contracts/broker/contracts/src/validators/GodsUnchainedValidator.sol:
```diff
@@ -0,0 +1,64 @@
+
+        // Decode and validate asset data.
+        address token = assetData.readAddress(16);
+        require(token == address(GODS_UNCHAINED), "TOKEN_ADDRESS_MISMATCH");
```
I think we're moving towards something like this now
```suggestion
        require(token == address(GODS_UNCHAINED), "GodsUnchainedValidator/TOKEN_ADDRESS_MISMATCH");
```

abandeali1 commented:
Another thought - if this contract approves the `StakingProxy` to spend it's WETH balance, then filling WETH denominated orders wouldn't require ETH to be sent with the transaction. This seems like it could be a pretty large UX improvement.

mzhu25 commented on contracts/broker/contracts/src/Broker.sol:
```diff
@@ -0,0 +1,253 @@
+        uint256[] calldata amounts,
+        bytes calldata data
+    )
+        external
```
should we check that `msg.sender == address(_EXCHANGE)` in the fallback while we're at it?

mzhu25 replied:
or I guess if we go with the WETH wrapping/unwrapping approach, `msg.sender == _WETH`

mzhu25 commented on contracts/broker/contracts/src/Broker.sol:
```diff
@@ -0,0 +1,253 @@
+    /// @param fillFunctionSelector The selector for either `fillOrder` or `fillOrKillOrder`.
+    /// @return fillResults Amounts filled and fees paid by the maker and taker.
+    function brokerTrade(
+        bytes[] memory brokeredAssets,
```
yeah, we can encode the token address in the 1155 asset data at the cost of a little generality

mzhu25 commented:
> Another thought - if this contract approves the `StakingProxy` to spend it's WETH balance, then filling WETH denominated orders wouldn't require ETH to be sent with the transaction. This seems like it could be a pretty large UX improvement.

hm, that's an interesting thought... I guess this would look like:
- Broker wraps `msg.value` into WETH, if any ETH is provided in the call
- All protocol fees are paid in WETH (either supplied by the taker or acquired from the maker during settlement)
- And at the end: 
```
for (uint256 i = 0; i < orders.length; i++) {
    if (orders[i].makerAssetData != WETH) {
        orders[i].makerAssetData.transferOut(fillResults[i].makerAssetFilledAmount);
    }
}
ETHER_TOKEN.withdraw(remainingWethBalance);
msg.sender.transfer(remainingWethBalance);
```

abandeali1 commented:
I don't think you'd actually need to do any wrapping/unwrapping. The Broker receives WETH as part of the trade and then uses that to pay protocol fees automatically if `msg.value` is 0. The only change we'd really need to make is subtract `protocolFeePaid` in the `transferOut` call when `makerAssetData == wethAssetData && msg.value == 0`.

mzhu25 commented:
> I don't think you'd actually need to do any wrapping/unwrapping. The Broker receives WETH as part of the trade and then uses that to pay protocol fees automatically if `msg.value` is 0. The only change we'd really need to make is subtract `protocolFeePaid` in the `transferOut` call when `makerAssetData == wethAssetData && msg.value == 0`.

This would cause problems if `msg.value = 1` or e.g. if the `makerAssetAmount` of WETH is not enough to cover all the protocol fees in the transaction and the taker needs to supply some ETH to make up the difference

abandeali1 commented:
True, but nothing is stopping anyone from including ETH in the transaction if the `makerAssetAmount < protocolFee`. Also `msg.value = 1` already breaks protocol fees (`msg.value == protocolFee || msg.value == 0` must be true).

mzhu25 commented:
Not sure if I follow –– If `0 < msg.value < protocolFee` the Exchange/StakingProxy will fall back to paying the protocol fee in WETH. So if this is the case and `makerAssetData` for a Broker order is WETH, we'd need to subtract `protocolFeePaid` from the `transferOut` amount even though `msg.value > 0`.

abandeali1 commented:
Nope, `msg.value` must be exactly 0 to pay the protocol fee in WETH. See https://github.com/0xProject/0x-monorepo/blob/development/contracts/staking/contracts/src/fees/MixinExchangeFees.sol#L58 and https://github.com/0xProject/0x-monorepo/blob/development/contracts/staking/contracts/src/fees/MixinExchangeFees.sol#L165.

abandeali1 commented:
Nvm you're right. I was thinking about how the `StakingProxy` handles receiving protocol fees, not how the `Exchange` handles sending them to the `StakingProxy`. 

Wrapping/unwrapping seems like the most logical approach to handle all of the edge cases then. Pretty much exactly what the `Forwarder` does.

mzhu25 commented on contracts/broker/contracts/src/Broker.sol:
```diff
@@ -0,0 +1,253 @@
+
+        // Clear storage
+        delete _cachedAssetData;
+        _cacheIndex = 0;
```
We do in the case that the order's `takerAssetAmount > 1`.

mzhu25 commented:
I guess at that point we might as well approve the ERC20Proxy as well, as this would enable WETH taker fees and filling orders with `takerAssetData = WETH`

dorothy-zbornak commented on contracts/broker/contracts/src/Broker.sol:
```diff
@@ -0,0 +1,253 @@
+    /// @param to This should be the maker of the order.
+    /// @param amounts Should be an array of just one `uint256`, specifying the amount of the brokered assets to transfer.
+    /// @param data Encodes the validator contract address and any auxiliary data it needs for property validation.
+    function safeBatchTransferFrom(
```
OK, so here is my admittedly kind of lame attack vector for this and why we should probably be checking that `msg.sender == ERC1155Proxy`:

- Maker creates one of these orders with `makerFeeAssetData = ERC1155(...)` and `feeRecipient == maker`.
- Taker is like, w/e not my fee not problem so I'll fill this order.
- Taker is also like, my sloppy ass will pass in more ERC721s than the fill/order wants because the broker shouldn't use them anyway.
- Exchange fills taker -> maker, which leaves behind the excess ERC721s "in" the broker contract.
- Exchange then fills maker fee, which is an 1155 so it calls `onERC1155BatchReceived()` on `maker` which, surprise, is a smart contract.
- In `onERC1155BatchReceived()` the maker calls `Broker.safeBatchTransferFrom(address(broker), maker, [], [1], hex"0x")`
- Maker walks away with an extra ERC721.

mzhu25 replied:
This is actually the scenario that the `if (from != address(this))` protects against. But yeah, I'm going to add a `msg.sender == ERC1155Proxy` check just in case we're missing an edge case

mzhu25 commented:
Added forwarder-like affiliate fees and WETH support
Refactored the Forwarder a good deal, most notably moved LibAssetDataTransfer and MixinWeth into @0x/contracts-extensions so that they can be reused between the Broker and Forwarder. (Also I deleted everything that was previously in the extensions🚶)

abandeali1 commented on contracts/extensions/contracts/src/MixinWethUtils.sol:
```diff
@@ -80,37 +112,24 @@ contract MixinWeth is
-    /// @param wethSpent Amount of WETH spent when filling orders.
+    /// @dev Unwraps and refunds WETH to msg.sender.
+    /// @param refundAmount Amount of WETH balance to refund.
     function _transferEthRefund(
```
Nit: now that this is in a more generic package, it probably makes sense to give this a more generic name. The first name that comes to mind is `withdrawAndTransferEth`.
